xds/src/main/java/io/grpc/xds/internal/sds/DynamicSslContextProvider.java

```diff
@@ -0,0 +1,141 @@
+      CertificateValidationContext certificateValidationContext)
+      throws CertificateException, IOException, CertStoreException;
+
+  // this gets called only when requested secrets are ready...
+  protected final void updateSslContext() {
+    try {
+      CertificateValidationContext localCertValidationContext =
+          generateCertificateValidationContext();
+      SslContextBuilder sslContextBuilder = getSslContextBuilder(localCertValidationContext);
+      CommonTlsContext commonTlsContext = getCommonTlsContext();
+      if (commonTlsContext != null && commonTlsContext.getAlpnProtocolsCount() > 0) {
+        List<String> alpnList = commonTlsContext.getAlpnProtocolsList();
+        ApplicationProtocolConfig apn =
+            new ApplicationProtocolConfig(
+                ApplicationProtocolConfig.Protocol.ALPN,
+                ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
+                alpnList);
+        sslContextBuilder.applicationProtocolConfig(apn);
+      }
+      List<Callback> pendingCallbacksCopy = null;
+      SslContext sslContextCopy = null;
+      synchronized (pendingCallbacks) {
+        sslContext = sslContextBuilder.build();
+        sslContextCopy = sslContext;
+        pendingCallbacksCopy = clonePendingCallbacksAndClear();
+      }
+      makePendingCallbacks(sslContextCopy, pendingCallbacksCopy);
+    } catch (Exception e) {
+      onError(Status.fromThrowable(e));
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected final void callPerformCallback(
+          Callback callback, final SslContext sslContextCopy) {
+    performCallback(
+        new SslContextGetter() {
+          @Override
+          public SslContext get() {
+            return sslContextCopy;
+          }
+        },
+        callback
+    );
+  }
+
+  @Override
+  public final void addCallback(Callback callback) {
+    checkNotNull(callback, "callback");
+    // if there is a computed sslContext just send it
+    SslContext sslContextCopy = null;
+    synchronized (pendingCallbacks) {
+      if (sslContext != null) {
+        sslContextCopy = sslContext;
+      } else {
+        pendingCallbacks.add(callback);
+      }
+    }
+    if (sslContextCopy != null) {
+      callPerformCallback(callback, sslContextCopy);
+    }
+  }
+
+  private final void makePendingCallbacks(
+      SslContext sslContextCopy, List<Callback> pendingCallbacksCopy) {
+    for (Callback callback : pendingCallbacksCopy) {
+      callPerformCallback(callback, sslContextCopy);
+    }
+  }
+
+  /** Propagates error to all the callback receivers. */
+  public final void onError(Status error) {
+    for (Callback callback : clonePendingCallbacksAndClear()) {
+      callback.onException(error.asException());
+    }
+  }
+
+  private List<Callback> clonePendingCallbacksAndClear() {
+    List<Callback> copy = ImmutableList.copyOf(pendingCallbacks);
```

Comment:
This method still needs `synchronized (pendingCallbacks)` because `onError()` is calling it, and need be synchronized with `addCallback()`.

Reply:
done